MFLT-7240 Write documentation for Linux coredump feature (#312)
memfault/memfault-docs: docs/linux/linux-coredumps.mdx

---
id: linux-coredumps
title: Linux Coredumps
sidebar_label: Coredumps
---

import { ProjectMagicLink } from "@site/src/components/ProjectMagicLink";

> ## Introduction

Author: TODO: a screenshot would be nice.

issues][docs-issue-management] as well as providing [metrics on
issues][docs-issue-charts] and monitoring via
[notifications][docs-notifications], granting you a clear image of how your
fleet is behaving, as well as tight control over the success of your [OTA
updates][docs-platform-ota].

[docs-issue-management]: /docs/platform/issue-management
[docs-issue-charts]: /docs/platform/charts/#issue-charts
[docs-notifications]: /docs/platform/notifications
[docs-platform-ota]: /docs/platform/ota

:::tip

Keep [`meta-memfault-example`][source-yocto-example] open as a reference
implementation. Your integration should look similar to it once you're done
following the steps in this tutorial.

:::

## Prerequisites

### The `memfaultd` daemon, built with `coredump_plugin`

Follow the [getting-started guide][docs-linux-getting-started] to learn how to
set this up for your device. A key function of `memfaultd` is to preprocess and
upload coredumps to the Memfault platform. It does this via `coredump_plugin`.

`coredump_plugin` is enabled by default. Read more on [how to configure which
plugins `memfaultd` builds with][docs-linux-control-plugins].

### Linux kernel configuration

> Build your Linux kernel with the following options enabled:

Reviewer: This makes it sound like there's an action to take, while I think it's highly likely to already be configured like this.

I'd rephrase to say something like

> Ensure that your Linux kernel with the following options enabled:

I asked [here](https://memfault.slack.com/archives/C037Z581MH9/p1664197171609859) how to check the existing defconfig/kconfig options:

Check `tmp/work/[machinename]/[kernelpackage]/[version]/build/.config` for your project's kernel configuration. For example: `tmp/work/qemuarm64-poky-linux/linux-yocto/5.15.62+gitAUTOINC+59c8898d45_7cb30c5e95-r0/linux-qemuarm64-standard-build/.config`

Author: Awesome, thanks!

- `CONFIG_COREDUMP=y` in order to enable coredump creation by the kernel
- `CONFIG_CORE_DUMP_DEFAULT_ELF_HEADERS=y` in order to enable default mappings.
  for processes. Specifically, Memfault requires:
  - `bit 0`: dump anonymous private mappings,
  - `bit 1`: dump anonymous shared mappings, and
  - `bit 4`: (available since Linux 2.6.24) dump ELF headers.

Read [`man core`][man-core] in order to learn more about these settings. We
recommend reading the entirety of this `man` page if you're in need of basic
understanding of how coredumps work in Linux.

### `core_pattern` and `core_pipe_limit`

The following files need to be owned by `memfaultd` completely.

- `/proc/sys/kernel/core_pattern` (see [`man core`][man-core])
- `/proc/sys/kernel/core_pipe_limit` (see [`man core`][man-core])

Make sure your Linux image does not contain any other services that may write to
these files. To do this, check for the inclusion of other coredump handlers
(although uncommon in an embedded system, this may be Ubuntu's
[Apport][apport]), and check your `/etc/sysctl.d` drop-in directory for anything
other than Memfault that may be setting `kernel.core_pattern` or

> `kernel.core_pipe_limit`.

Reviewer: Systemd also comes with coredump handling. That's more likely to be in the way than apport I think.

Looking at [this recipe for systemd](https://git.yoctoproject.org/poky/plain/meta/recipes-core/systemd/systemd_251.4.bb) there's a `coredump` option. Probably need to mention that it should not be used.

To check whether you've succeeded in letting `memfaultd` take care of these
files, confirm that the contents of the `core_pattern` file reference the

> `memfault-core-handler` binary.

Reviewer: The pattern is only written if memfaultd has data collection enabled. I'd mention that here.

## Uploading debugging symbols

> In order to process coredumps from your devices, you'll need to upload debugging

Reviewer: ```suggestion
In order for Memfault to process coredumps from your devices, you'll need to upload debugging
```

symbols for each of the binaries in your Linux image. We recommend doing this as
part of your build process, either locally or ideally as part of continuous
integration, in order to keep Memfault up to date with your build's newest
debugging symbols.

In order to facilitate this, we've added a Yocto-specific helper subcommand to
the [Memfault CLI][docs-memfault-cli]: `memfault upload-yocto-symbols`. Here's
an example invocation:

```bash
$ memfault \
    --org $YOUR_ORGANIZATION_SLUG \
    --org-token $ORGANIZATION_AUTH_TOKEN \
    --project $YOUR_PROJECT_SLUG upload-yocto-symbols \
    --manifest deploy/images/raspberrypi3/base-image-raspberrypi3.manifest
```

Read docs on [CI and authentication][docs-ci-auth] in order to obtain
credentials that you can use in your build environment or in continuous
integration.

Yocto's `do_rootfs` will produce a `.manifest` file (see [documentation on
`IMAGE_MANIFEST`][yocto-image-manifest]) which lists all the installed packages
that make up the image. The `memfault upload-yocto-symbols` commands makes use

> of it in order to find your build's debugging symbols.

Reviewer: In the help I had added:

>     The .manifest file is expected to reside in the location where Yocto produced the file.

to ensure hint that people shouldn't move the .manifest.

Do you think we should mention that here too?

If you're not using Yocto, you can replicate its behavior by using our [REST API

> to upload symbols][api-docs-upload-symbols].

Reviewer: OT: wonder if we should add a "generic" `upload-elf-symbols` command that takes a (directory with) ELF file(s), rather than pointing people to eject out of the CLI tool and use the REST API. Note this command would be a synonym for the current `upload-aosp-symbols` command.

<p>
  You can always upload symbol files individually using the Memfault Web
  Application. Open <b>Software -> Symbol Files</b> and click on{" "}
  <b>Upload Symbol File</b>, or simply{" "}
  <ProjectMagicLink path="/symbol-files?new">
    click on this deep link to the app
  </ProjectMagicLink>
  .
</p>

> 

Reviewer: People who've used Yocto + GDB may wonder whether they need to opt-in to any of the Yocto-provided debug options, like `IMAGE_GEN_DEBUGFS`, `DEBUG_BUILD`, `dbg-pkgs`, etc. Should we call out that none of these are needed and no additional changes to the device's image are required?

## Set `enable_data_collection`

By default, `enable_data_collection` is `false` (see the [default
configuration][source-builtin-conf]). This is to enable asking end users for
consent before collecting or transmitting any data to Memfault services.

Once the end user has given their consent, you can enable data collection like
so:

```shell
$ memfaultd --enable-data-collection
```

To disable it:

```shell
$ memfaultd --disable-data-collection
```

The `memfaultd` service will restart automatically whenever you run either of

> those commands.

Reviewer: Actually, only if it changed/toggled the setting:

https://github.com/memfault/memfault-linux-sdk-internal/blob/cffc3068dad3f4cadbe239cb4b29a5853f66d9f2/meta-memfault/recipes-memfault/memfaultd/files/memfaultd/src/memfaultd.c#L160-L161

Take a look at the [`/etc/memfaultd.conf` reference][docs-configuration] for
more information.

[source-builtin-conf]:
  https://github.com/memfault/memfault-linux-sdk/blob/-/meta-memfault/recipes-memfault/memfaultd/files/memfaultd/builtin.conf

## Configuring what's included in a coredump

Currently, Memfault support for coredumps does not include custom filtering or

> selection of processes, memory regions or variables. However, the kernel does

Reviewer: Should we mention that this will be added in a future SDK?

I'd rather not have people go all-in on `madvise` / etc. because those controls are super crude.
On top of that, once we add filtering ourselves, it will be counter-productive if those controls are *also* used by customers (will be harder to debug problems in the filtering logic).

Author: Changed to:

```
Currently, Memfault support for coredumps does not include custom filtering or
selection of processes, memory regions or variables. These features are planned
for a future release.

However, the kernel does provide a certain degree of configurability. We'd
recommend waiting for a release of the Memfault Linux SDK with first-class
support for coredump filtering, but if your team needs it now:
```

provide a certain degree of configurability, and you can use it in your project:

- Call `prctl(PR_SET_DUMPABLE, 0)` in order to declare a process as
  non-dumpable. By default, processes _are_ dumpable. See
  [`man prctl`][man-prctl] for more information.
- Set the appropiate flags in `/proc/[pid]/coredump_filter` (see
  [`man core`][man-core]).
  - Note that in order to work properly with Memfault, `coredump_filter` needs
    to include at least bits `0`, `1` and `4` (see [`man core`][man-core]).
- Call `madvise(ptr, bytes, MADV_DONTDUMP)` (with `MADV_DONTDUMP` or
  `MADV_DODUMP`) in order to declare certain memory regions as dumpable (or
  non-dumpable). The effects of this syscall take precedence over the settings
  in the `coredump_filter` for this file (see [`man madvise`][man-madvise] for
  more information).

[man-prctl]: https://man7.org/linux/man-pages/man2/prctl.2.html
[man-madvise]: https://man7.org/linux/man-pages/man2/madvise.2.html

## Testing your integration

In order to test your integration, you'll need to make your device produce a
coredump for a process, for example, by sending `SIGABRT` to it:

```bash
$ kill -s SIGABRT $YOUR_PROCESS_PID
```

> Additionally, you might want to check that the `core_pattern` file is being set

Reviewer: Fine for now, but would probably move this piece to a "Troubleshooting" section on this page.

---
id: linux-coredumps
title: Linux Coredumps
sidebar_label: Coredumps
---

import { ProjectMagicLink } from "@site/src/components/ProjectMagicLink";

## Introduction

From [`man core`][man-core]:

> The default action of certain signals is to cause a process to terminate and
> produce a core dump file, a file containing an image of the process's memory
> at the time of termination. This image can be used in a debugger (e.g.,
> gdb(1)) to inspect the state of the program at the time that it terminated. A
> list of the signals which cause a process to dump core can be found in

> > signal(7).

Reviewer 2: I'd move this section below our own intro paragraph.

By enabling coredump support with Memfault, the `memfaultd` daemon will
automatically collect, preprocess and upload coredumps from your devices and

> send them to the Memfault platform. In turn, the Memfault platform acts as a

Reviewer 2: send them immediately? regularly?

Author: IMO to early in the doc for this detail. Will consider mentioning this later.

remote analyzer for your coredump files and (using your project's debugging
symbols) is able to display a rich view of a coredump, displaying backtraces for
all threads, and allowing you to inspect state in full detail.

Additionally, Memfault takes care of [managing and deduplicating
issues][docs-issue-management] as well as providing [metrics on
issues][docs-issue-charts] and monitoring via
[notifications][docs-notifications], granting you a clear image of how your
fleet is behaving, as well as tight control over the success of your [OTA

> updates][docs-platform-ota].

Reviewer 2: I like how we continue to push the docs towards describing the value gain beyond the plain feature description 👍

<ImageFigure

>   src="/img/docs/linux/linux-issue-detail-sleep-sigabrt.png"

Reviewer 2: if we could have a screenshot that with more traces and devices, we could use the caption to hint at the value of fleet-wide coredump analysis, e.g. "A Linux coredump as it occurred hundreds of times already on 52 different devices in the past 24 hours. Memfault represents actionable details as a deduplicated Issue"

  alt="A Linux coredump as an Issue on the Memfault Web App"
  title="Linux coredump on Memfault"
>
  A Linux coredump as an Issue on the Memfault Web App.
</ImageFigure>

[docs-issue-management]: /docs/platform/issue-management
[docs-issue-charts]: /docs/platform/charts/#issue-charts
[docs-notifications]: /docs/platform/notifications
[docs-platform-ota]: /docs/platform/ota

:::tip

Keep [`meta-memfault-example`][source-yocto-example] open as a reference
implementation. Your integration should look similar to it once you're done
following the steps in this tutorial.

:::

## Prerequisites

### The `memfaultd` daemon, built with `coredump_plugin`

Follow the [getting-started guide][docs-linux-getting-started] to learn how to
set this up for your device. A key function of `memfaultd` is to preprocess and

> upload coredumps to the Memfault platform. It does this via `coredump_plugin`.

Reviewer 2: ```suggestion
upload coredumps to the Memfault platform. It does this via its `coredump_plugin`.
```

`coredump_plugin` is enabled by default. Read more on [how to configure which
plugins `memfaultd` builds with][docs-linux-control-plugins].

### Linux kernel configuration

Ensure that your Linux kernel is built with the following options enabled:

- `CONFIG_COREDUMP=y` in order to enable coredump creation by the kernel
- `CONFIG_CORE_DUMP_DEFAULT_ELF_HEADERS=y` in order to enable default mappings.
  for processes. Specifically, Memfault requires:
  - `bit 0`: dump anonymous private mappings,
  - `bit 1`: dump anonymous shared mappings, and

>   - `bit 4`: (available since Linux 2.6.24) dump ELF headers.

Reviewer 2: Where do these bits come from? Is this something the user can configure beyond a plain `y`? Is this configured elsewhere?

Author: Two paragraphs later:

> Read [`man core`][man-core] in order to learn more about these settings. We
recommend reading the entirety of this `man` page if you're in need of basic
understanding of how coredumps work in Linux.

I'm moving that up to come right after this list.

In order to confirm that your kernel is correctly configured, check out your
project's configuration file in
`tmp/work/[machine]/[kernelpackage]/[version]/build/.config`. For example:

```
tmp/work/raspberrypi3-poky-linux/linux-yocto/5.15.62+gitAUTOINC+59c8898d45_7cb30c5e95-r0/linux-raspberrypi3-standard-build/.config`
```

Read [`man core`][man-core] in order to learn more about these settings. We
recommend reading the entirety of this `man` page if you're in need of basic
understanding of how coredumps work in Linux.

### `core_pattern` and `core_pipe_limit`

> The following files need to be owned by `memfaultd` completely.

Reviewer 2: The "owned" term sounds strange to me here. It unclear to me if these files normally exist or if they are also created by `memfaultd`

```suggestion
The following files will be modified by `memfaultd` at runtime and must not be written to by any other process.
```

- `/proc/sys/kernel/core_pattern` (see [`man core`][man-core])
- `/proc/sys/kernel/core_pipe_limit` (see [`man core`][man-core])

Make sure your Linux image does not contain any other services that may write to
these files. To do this, check for the inclusion of other coredump handlers such
as SystemD (which can [act as a coredump handler when built with
`-Dcoredump=true`][systemd-as-handler]), and check your `/etc/sysctl.d` drop-in
directory for anything other than Memfault that may be setting
`kernel.core_pattern` or `kernel.core_pipe_limit`.

To check whether you've succeeded in letting `memfaultd` take care of these

> files, confirm that the contents of the `core_pattern` file reference the

Reviewer 2: Do we later discuss how one confirms a correct setup at runtime? If so, maybe hint at this here?

Author: :+1: Added a link

`memfault-core-handler` binary.

> Note that `memfaultd` only sets `core_pattern`
> [if data collection is enabled](#set-enable_data_collection).

Reviewer 2: ```suggestion
Note that `memfaultd` respects privacy settings only sets `core_pattern`
[if data collection is enabled](#set-enable_data_collection) (at runtime).
```

How do we handle the case when a user revokes data collection?

Author: `journalctl` will log "won't post this corefile. Data collection disabled" or similar. I'll comment on `journalctl` logs later in the doc.

[systemd-as-handler]:
  https://git.yoctoproject.org/poky/plain/meta/recipes-core/systemd/systemd_251.4.bb

## Upload debugging symbols

> In order for Memfault to process coredumps from your devices, you'll need to
> upload debugging symbols for each of the binaries in your Linux image. We

Reviewer 2: So far, the reader was purely thinking about software on the device. I am suggesting a few words here to ease the transition to the system-level thinking.

Ideally, we want a diagram here but I think that can wait.

```suggestion
After a coredump was collected on the device, the `memfaultd` will upload it to the Memfault Web Application. To allow the Memfault Web Application to reconstruct all details from such coredumps, you'll need to
upload debugging symbols for each of the binaries in your Linux image. We
```

> recommend doing this as part of your build process (since this needs to be done

Reviewer 2: The motivation is unclear, we need to do this for each new build of the software (hasn't been said so far). To reduce the risk of ever missing any software version, we recommend doing that in an automated fashion.

Maybe also worth calling out: It doesn't harm if you attempt to upload the same symbols multiple times. Memfault will detect this and the CLI will only upload what's new and truly needed.

credentials that you can use in your build environment or in continuous
integration.

Yocto's `do_rootfs` will produce a `.manifest` file (see [documentation on
`IMAGE_MANIFEST`][yocto-image-manifest]) which lists all the installed packages
that make up the image. The `memfault upload-yocto-symbols` commands makes use
of it in order to find your build's debugging symbols. The `.manifest` file is
expected to reside in the location where Yocto produced it.

If you're not using Yocto, you can replicate its behavior by using our [REST API
to upload symbols][api-docs-upload-symbols].

<p>
  You can always upload symbol files individually using the Memfault Web
  Application. Open <b>Software -> Symbol Files</b> and click on{" "}
  <b>Upload Symbol File</b>, or simply{" "}
  <ProjectMagicLink path="/symbol-files?new">
    click on this deep link to the app
  </ProjectMagicLink>
  .
</p>

:::tip

Note that in order to upload debugging symbols it's not necessary to use any of
Yocto's built-in debugging features such as `IMAGE_GEN_DEBUGFS`, `DEBUG_BUILD`
or `dgb-pkgs`. The defaults can be used and will work with
`memfault upload-yocto-symbols`.

:::

## Set `enable_data_collection`

> By default, `enable_data_collection` is `false` (see the [default
> configuration][source-builtin-conf]). This is to enable asking end users for

Reviewer 2: ```suggestion
By default, `enable_data_collection` is `false` (see the [default
configuration][source-builtin-conf] to change this). This is to enable asking end users for
```

Author: They can't change it there. They would have to change it in their own `memfaultd.conf`, and IMO changing the default (setting it to `true`) shouldn't be recommended, since that would skip asking users for consent.

  [`man core`][man-core]).
  - Note that in order to work properly with Memfault, `coredump_filter` needs
    to include at least bits `0`, `1` and `4` (see [`man core`][man-core]).
- Call `madvise(ptr, bytes, MADV_DONTDUMP)` (with `MADV_DONTDUMP` or
  `MADV_DODUMP`) in order to declare certain memory regions as dumpable (or
  non-dumpable). The effects of this syscall take precedence over the settings
  in the `coredump_filter` for this file (see [`man madvise`][man-madvise] for
  more information).

[man-prctl]: https://man7.org/linux/man-pages/man2/prctl.2.html
[man-madvise]: https://man7.org/linux/man-pages/man2/madvise.2.html

## Test your integration

In order to test your integration, you'll need to make your device produce a
coredump for a process, for example, by sending `SIGABRT` to it:

```bash
$ kill -s SIGABRT $YOUR_PROCESS_PID
```

Additionally, you might want to check that the `core_pattern` file is being set
correctly by `memfaultd` when it starts. Check that the output of this command
includes `memfault-core-handler`:

```bash
$ cat /proc/sys/kernel/core_pattern
```

<p>
  If your integration is all set and you've{" "}

>   <a href="#set-enable_data_collection">enabled data collection</a>, you'll be

Reviewer 2: is there anything else they can already do on the device? e.g. logs of sorts?

Author: Added a mention of `journalctl --unit memfaultd`.

  able to see a new issue pop up in{" "}
  <ProjectMagicLink path="/issues?live">
    your project's Issues page
  </ProjectMagicLink>
  .

> </p>

Reviewer 2: nit: why is this not Markdown?

Author: Prettier fails to format `<ProjectMagicLink...` if I put it in the middle of the Markdown, removing necessary whitespace.

IMAGE_GEN_DEBUGFS = "1"
IMAGE_FSTYPES_DEBUGFS = "tar.bz2"
```

Note that `IMAGE_FSTYPES_DEBUGFS` must match the `IMAGE_FSTYPE` of the image you
pass to the command.

Here's an example `memfault upload-yocto-symbols` invocation:

```bash
$ source oe-init-build-env
$ memfault \
    --org $YOUR_ORGANIZATION_SLUG \
    --org-token $ORGANIZATION_AUTH_TOKEN \
    --project $YOUR_PROJECT_SLUG upload-yocto-symbols \
    --image tmp/deploy/images/raspberrypi3/base-image-raspberrypi3.tar.bz2
```

It's recommended that you run this command after having sourced the build
environment script using `source oe-init-build-env`.

<details>
<summary>
  If you can't do this, then you can pass the information needed from the build environment as
  command-line arguments:
</summary>

- `--eu-unstrip-path`: path to a local `eu-unstrip` binary from
  [elfutils][elfutils]. Note that one is available in
  `tmp/sysroot-components/x86_64/elfutils-native` if you've added
  `DEPENDS:append = " elfutils-native"` to your build.
- `--package-debug-split-style`: your project's `PACKAGE_DEBUG_SPLIT_STYLE`. In
  Poky, it defaults to `debug-with-srcpkg`. Read more about it [in the Yocto

>   reference][split-style].

Reviewer 2: this is super valuable detail, thank you!
